lib/mayaUsd/ufe/UsdUndoCreateStageWithNewLayerCommand.cpp

```diff
@@ -63,148 +62,124 @@ void UsdUndoCreateStageWithNewLayerCommand::execute()
-        if (MFAIL(status)) {
-            throw std::runtime_error("");
-        }
-
-        // Execute the operations.
-        status = _createProxyShapeDagMod.doIt();
-        if (MFAIL(status)) {
-            throw std::runtime_error("");
-        }
-        _success = true;
-
-        // Create a UFE scene item for the newly created mayaUsdProxyShape.
-        Ufe::Path proxyShapeUfePath = MayaUsd::ufe::dagPathToUfe(proxyShapeDagPath);
-        _insertedChild = Ufe::Hierarchy::createItem(proxyShapeUfePath);
-
-        // Refresh the cache of the stage map.
-        // When creating the proxy shape, the stage map gets dirtied and cleaned. Afterwards, the
-        // proxy shape is renamed. The stage map does not observe the Maya data model, so renaming
-        // does not dirty the stage map again. Thus, the cache is in an invalid state, where it
-        // contains the path of the proxy shape before it was renamed. Calling getProxyShape()
-        // refreshes the cache. See comments within UsdStageMap::proxyShape() for more details.
-        getProxyShape(proxyShapeUfePath);
+        OpUndoItemRecorder undoRecorder(_undoItemList);
+        success = executeWithinUndoRecorder();
     } catch (const std::exception&) {
-        if (createTransformSuccess) {
-            _createTransformDagMod.undoIt();
-        }
+        _undoItemList.undo();
+        throw;
+    }
+
+    if (!success) {
+        _undoItemList.undo();
     }
 }
 
-void UsdUndoCreateStageWithNewLayerCommand::undo()
+bool UsdUndoCreateStageWithNewLayerCommand::executeWithinUndoRecorder()
 {
-    if (_success) {
-        _createProxyShapeDagMod.undoIt();
-        _createTransformDagMod.undoIt();
-    }
+    // Get a MObject from the parent scene item.
+    // Note: If and only if the parent is the world node, MDagPath::transform() will set status
+    // to kInvalidParameter. In this case MObject::kNullObj is returned, which is a valid parent
+    // object. Thus, kInvalidParameter will not be treated as a failure.
+    MStatus  status;
+    MDagPath parentDagPath = MayaUsd::ufe::ufeToDagPath(_parentItem->path());
+    MObject  parentObject = parentDagPath.transform(&status);
+    if (status != MStatus::kInvalidParameter && MFAIL(status))
+        return false;
+
+    MDagModifier& dagMod = MDagModifierUndoItem::create("Create stage with new Layer");
+
+    // Create a transform node.
+    // Note: It would be possible to create the transform and the proxy shape in one doIt() call
+    // of a single MDagModifier. However, doing so causes notifications to be sent in a
+    // different order, which triggers a `TF_VERIFY(g_StageMap.isDirty())` in
+    // StagesSubject::onStageSet(). Using a separate MDagModifier to create the transform seems
+    // more robust and avoids triggering the TF_VERIFY.
+    MObject transformObj;
+    transformObj = dagMod.createNode("transform", parentObject);
+    if (transformObj.isNull())
+        return false;
+
+    if (!dagMod.doIt())
```

Comment:
The boolean conversion of MStatus if equivalent to checking if it failed. Like you said, it makes the code slightly more readable in general than having a macro (especially since the macro is not named in the form of a question, not even a verb, IMO it would have read better if the macro was MFAILED. In any case, converting to bool is pretty standard for functions return MStatus unless you want to check for a specific error value.

Reply:
I see, makes sense!